### PR TITLE
gpsd: patch dependent packages to be compatible with version in nixpkgs

### PR DESCRIPTION
**Pull request by hjones2199**
###### Motivation for this change
gpsd 3.23.1 update (#139794) broke compilation of packages by updating the names of c macros (particularly STATUS_NO_FIX -> STATUS_UNK). Auto-patching them with substituteInPlace until upstream supports the new macros solves the issue. The KDE packages that failed to build only transitively depend on GPSD via marble, so that fix should allow the rest of the kde packages (calligra, kreport etc) to build successfully. 

The only package that still seems to be broken is timedoctor, i'm not sure how to proceed on that one since its extracted from an appimage.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 140523` at 55bf1692 run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>7 packages skipped due to time constraints:</summary>
  <ul>
    <li>calligra</li>
    <li>digikam</li>
    <li>kexi</li>
    <li>libsForQt512.kreport</li>
    <li>libsForQt512.marble</li>
    <li>libsForQt514.kreport</li>
    <li>libsForQt514.marble</li>
  </ul>
</details>
<details>
  <summary>4 packages built successfully:</summary>
  <ul>
    <li>foxtrotgps</li>
    <li>libsForQt5.kreport (libsForQt515.kreport ,plasma5Packages.kreport)</li>
    <li>libsForQt5.marble (libsForQt515.marble ,plasma5Packages.marble)</li>
    <li>zombietrackergps</li>
  </ul>
</details>
<details>
  <summary>3 suggestions:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/unclear-gpl.md">unclear-gpl</a>

`lgpl21` is a deprecated license, please check if project uses `lgpl21Plus` or `lgpl21Only` and change `meta.license` accordingly.

Near pkgs/applications/kde/marble.nix:13:5:
```
   |
13 |     license = with lib.licenses; [ lgpl21 gpl3 ];
   |     ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/maintainers-missing.md">maintainers-missing</a>

Package does not have a maintainer. Consider adding yourself?

Near pkgs/applications/kde/default.nix:51:9:
```
   |
51 |         meta =
   |         ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/unclear-gpl.md">unclear-gpl</a>

`gpl3` is a deprecated license, please check if project uses `gpl3Plus` or `gpl3Only` and change `meta.license` accordingly.

Near pkgs/applications/kde/marble.nix:13:5:
```
   |
13 |     license = with lib.licenses; [ lgpl21 gpl3 ];
   |     ^
```
</li>
  </ul>
</details>

File: pkgs/applications/kde/marble.nix
Comment (timokau) [inline pkgs/applications/kde/marble.nix:21]: Please add a link to the upstream MR @CrystalGamma mentioned in #139794 (https://invent.kde.org/education/marble/-/merge_requests/57). I would prefer to use the patch from that MR directly, but if you prefer `substituteInPlace` that is also okay.

File: pkgs/applications/misc/foxtrotgps/default.nix
Comment (timokau) [inline pkgs/applications/misc/foxtrotgps/default.nix:46]: Please add more information that makes it easier to determine if this can be removed. Assuming the build currently fails without the change, I suggest

```suggestion
  # Patch for compatibility with gpsd 3.23.1. This was added for foxtrotgps
  # 1.2.2+331. The command can be removed if the build of a newer version
  # succeeds without it.
```

**Review by timokau — CHANGES_REQUESTED** [review]
Thank you for working on this! I have some minor suggestions.

**Comment by timokau** [comment]
CC @kamadorueda regarding `timedoctor`. The package is currently broken because it is incompatible with gpsd 3.23.1.  If you have a way of contacting upstream, it may be a good idea to let them know.

File: pkgs/applications/misc/foxtrotgps/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/misc/foxtrotgps/default.nix:47]:
```suggestion
  postPatch = ''
```

File: pkgs/applications/kde/marble.nix
Comment (SuperSandro2000) [inline pkgs/applications/kde/marble.nix:22]:
```suggestion
  postPatch = ''
```
If you see any other package that uses prePatch for this instead of post, please send a PR in. This is wrong and breaks applying patches.

**Comment by kamadorueda** [comment]
> CC @kamadorueda regarding `timedoctor`. The package is currently broken because it is incompatible with gpsd 3.23.1. If you have a way of contacting upstream, it may be a good idea to let them know.

Sure, no problem, thanks for letting me know

File: pkgs/applications/misc/foxtrotgps/default.nix
Comment (hjones2199) [reply pkgs/applications/misc/foxtrotgps/default.nix]: Changed patching strategy entirely for marble (fetching from upstream git), and changed the foxtrotgps one to postPatch.

**Review by timokau — APPROVED** [review]
Thank you!

Edit: Waiting for ofBorgs approval.

**Comment by hjones2199** [comment]
From looking at last successful build of calligra, (https://hydra.nixos.org/build/149442341) it seems it was broken before the gpsd upgrade and for some reason not marked as broken?

**Comment by austinbutler** [comment]
This fixed the build of Digikam for me.

**Pull request merged**
